server/webhook.go

```diff
@@ -537,6 +543,44 @@ func (s *Server) processPushEvent(ctx context.Context, repositoryList []*api.Rep
 	return createdMessageList, nil
 }
 
+// Users may merge commits from other branches, and some of the commits merged in may already be merged into the main branch.
+// In that case, the commits in the push event contains files which are not added in this PR/MR.
+// We use the compare API to get the file diffs and filter files by the diffs.
+func (s *Server) filterFilesByCommitsDiff(ctx context.Context, repo *api.Repository, distinctFileList []vcs.DistinctFileItem, beforeCommit, afterCommit string) ([]vcs.DistinctFileItem, error) {
+	fileDiffList, err := vcs.Get(repo.VCS.Type, vcs.ProviderConfig{}).GetDiffFileList(
+		ctx,
+		common.OauthContext{
+			ClientID:     repo.VCS.ApplicationID,
+			ClientSecret: repo.VCS.Secret,
+			AccessToken:  repo.AccessToken,
+			RefreshToken: repo.RefreshToken,
+			Refresher:    s.refreshToken(ctx, repo.WebURL),
+		},
+		repo.VCS.InstanceURL,
+		repo.ExternalID,
+		beforeCommit,
+		afterCommit,
+	)
+	if err != nil {
+		return nil, errors.WithMessagef(err, "failed to get file diff list for repository %s", repo.ExternalID)
+	}
+	var filteredDistinctFileList []vcs.DistinctFileItem
+	for _, file := range distinctFileList {
+		for _, diff := range fileDiffList {
+			if file.FileName == diff.Path {
+				filteredDistinctFileList = append(filteredDistinctFileList, file)
+				break
+			}
+		}
+	}
```

Comment:
If we already have the list of files that's been changed... there seems no point in comparing with them? i.e. couldn't we just use the returned file list (`fileDiffList`) directly?

Reply:
Yes. It's the first comment here.
https://github.com/bytebase/bytebase/pull/3411#issuecomment-1317021201

Reply:
Ah, right, haha!